File: pgpainless-core/src/main/java/org/pgpainless/encryption_signing/BcHashContextSigner.java
// SPDX-FileCopyrightText: 2022 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.encryption_signing;

import java.security.MessageDigest;
import java.util.List;

import org.bouncycastle.openpgp.PGPException;
import org.bouncycastle.openpgp.PGPPrivateKey;
import org.bouncycastle.openpgp.PGPPublicKey;
import org.bouncycastle.openpgp.PGPSecretKey;
import org.bouncycastle.openpgp.PGPSecretKeyRing;
import org.bouncycastle.openpgp.PGPSignature;
import org.bouncycastle.openpgp.PGPSignatureGenerator;
import org.pgpainless.PGPainless;
import org.pgpainless.algorithm.SignatureType;
import org.pgpainless.key.info.KeyRingInfo;
import org.pgpainless.key.protection.SecretKeyRingProtector;
import org.pgpainless.key.protection.UnlockSecretKey;

import javax.annotation.Nonnull;

public class BcHashContextSigner {

    public static PGPSignature signHashContext(@Nonnull MessageDigest hashContext,
                                               @Nonnull SignatureType signatureType,
                                               @Nonnull PGPSecretKeyRing secretKeys,
                                               @Nonnull SecretKeyRingProtector protector)
            throws PGPException {
        KeyRingInfo info = PGPainless.inspectKeyRing(secretKeys);
        List<PGPPublicKey> signingSubkeyCandidates = info.getSigningSubkeys();
        PGPSecretKey signingKey = null;
        for (PGPPublicKey signingKeyCandidate : signingSubkeyCandidates) {
            signingKey = secretKeys.getSecretKey(signingKeyCandidate.getKeyID());
            if (signingKey != null) {
                break;
            }
        }
        if (signingKey == null) {
            throw new PGPException("Key does not contain suitable signing subkey.");
        }

        PGPPrivateKey privateKey = UnlockSecretKey.unlockSecretKey(signingKey, protector);
        return signHashContext(hashContext, signatureType, privateKey);
    }

    /**
     * Create an OpenPGP Signature over the given {@link MessageDigest} hash context.
     *
     * @param hashContext hash context
     * @param privateKey signing-capable key
     * @return signature
     * @throws PGPException in case of an OpenPGP error
     */
    static PGPSignature signHashContext(MessageDigest hashContext, SignatureType signatureType, PGPPrivateKey privateKey)
            throws PGPException {
        PGPSignatureGenerator sigGen = new PGPSignatureGenerator(
                new BcPGPHashContextContentSignerBuilder(hashContext)
        );

        sigGen.init(signatureType.getCode(), privateKey);
        return sigGen.generate();
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/encryption_signing/BcHashContextSigner.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.encryption_signing

import org.bouncycastle.openpgp.PGPException
import org.bouncycastle.openpgp.PGPSignatureGenerator
import org.bouncycastle.openpgp.api.OpenPGPKey
import org.bouncycastle.openpgp.api.OpenPGPSignature.OpenPGPDocumentSignature
import org.pgpainless.PGPainless
import org.pgpainless.algorithm.SignatureType
import org.pgpainless.key.protection.SecretKeyRingProtector
import org.pgpainless.key.protection.UnlockSecretKey
import java.security.MessageDigest

class BcHashContextSigner {

    companion object {
        @JvmStatic
        fun signHashContext(
            hashContext: MessageDigest,
            signatureType: SignatureType,
            secretKey: OpenPGPKey,
            protector: SecretKeyRingProtector
        ): OpenPGPDocumentSignature {
            val info = PGPainless.getInstance().inspect(secretKey)
            return info.signingSubkeys
                .mapNotNull { info.getSecretKey(it.keyIdentifier) }
                .firstOrNull()
                ?.let {
                    signHashContext(hashContext, signatureType, UnlockSecretKey.unlockSecretKey(it, protector))
                }
                ?: throw PGPException("Key does not contain suitable signing subkey.")
        }

        /**
         * Create an OpenPGP Signature over the given [MessageDigest] hash context.
         *
         * @param hashContext hash context
         * @param privateKey signing-capable key
         * @return signature
         * @throws PGPException in case of an OpenPGP error
         */
        @JvmStatic
        internal fun signHashContext(
            hashContext: MessageDigest,
            signatureType: SignatureType,
            privateKey: OpenPGPKey.OpenPGPPrivateKey
        ): OpenPGPDocumentSignature {
            return PGPSignatureGenerator(
                BcPGPHashContextContentSignerBuilder(hashContext),
                privateKey.keyPair.publicKey)
                .apply { init(signatureType.code, privateKey.keyPair.privateKey) }
                .generate()
                .let { OpenPGPDocumentSignature(it, privateKey.publicKey) }
        }
    }
}
